xADObjectPermissionEntry: New resource

claudiospizzi opened the pull request:
I've developed a new resource to control the access control entries of Active Directory objects. With this implementation, we can control the ace entries and ensure, that they should be present or absent.

My use case is the SQL Server AlwaysOn Availability Group Cluster, where we have to prestage a VCO and add full control permissions (aka GenericAll) to the CNO.

Fixes #57

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/powershell/xactivedirectory/186)
<!-- Reviewable:end -->

johlju commented:
@claudiospizzi Can you please rebase this PR so you can resolve the conflicts?

claudiospizzi commented:
@johlju Rebased and readme resolved.

johlju commented:
Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---

*[a discussion](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfIRi83t2cC3fETmem:-LBfIRi83t2cC3fETmen:bzds8tw) (no related file):*
I started reviewing this but soon realized that the Key's in the schema seems that they could become a problem in certain scenarios (I can be wrong). For me to get a bigger understanding for this resource, could you provide a few examples?
Currently there are no Examples folder, so maybe we can do something similar to https://github.com/PowerShell/SqlServerDsc/tree/dev/Examples? 

Can we start with three examples.

1. Give a prestaged VCO full control permissions (aka GenericAll) to the CNO.
2. Allow a group permission to create and delete  ('CreateChild,DeleteChild') computer objects  in an OU and all sub-OUs that may get created.
3. Allow a group permission to read and write ('ReadProperty,WriteProperty') all properties of computer objects in an OU and all sub-OUs that may get created

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 1 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfKBLM1iHXlhvJXCNi:-LBfKBLM1iHXlhvJXCNj:bmpqjg2) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L1)):*
> ```
> xADObjectPermissionEntry
> ```

Is it a special reason we should end the name with 'Entry'? Would it be misleading to call it just 'xADObjectPermission'?

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 5 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfUqiLC-d21git-10n:-LBfUqiLC-d21git-10o:b-sgc7m8) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L5)):*
> ```
> {
>     [Write, Description("Indicates if the access will be added (Present) or will be removed (Absent). Default is 'Present'."), ValueMap{"Present", "Absent"}, Values{"Present", "Absent"}] String Ensure;
>     [Key, Description("Active Directory path of the object, specified as a Distinguished Name.")] String Path;
> ```

Should we make the description more descriptive and say something about that this is the object (for example and OU?) that will get the permission?

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 6 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfKald4_65wNZ9w0SN:-LBfKald4_65wNZ9w0SO:b-tnw2k4) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L6)):*
> ```
> IdentityReference
> ```

Is there a special reason to add 'Reference' here? Can it only be assigned either SamAccountName, CommonName, UserPrincipalName, DisplayName or is it possible to provide any them? Thinking if we should call it just 'Identity', or if we should use any existing name that for example xUser uses?  

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 6 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfW4HU5tkO3HUvcEYm:-LBfW4HU5tkO3HUvcEYn:bnziti4) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L6)):*
> ```
>     [Write, Description("Indicates if the access will be added (Present) or will be removed (Absent). Default is 'Present'."), ValueMap{"Present", "Absent"}, Values{"Present", "Absent"}] String Ensure;
>     [Key, Description("Active Directory path of the object, specified as a Distinguished Name.")] String Path;
>     [Key, Description("Indicates the identity of the principal for the ace.")] String IdentityReference;
> ```

We should say in the description that this can be only be set to an user or group account.
We should replace the abbreviation 'ace' with something that it is this that gets 'permission'.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 7 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfLizg350lEka_feEi:-LBfLizg350lEka_feEj:b-bbc13a) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L7)):*
> ```
> ActiveDirectoryRights
> ```

Maybe call this property `Permission`?

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 9 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfRuZI7fGr7oUYo8fp:-LBfRuZI7fGr7oUYo8fq:b-tbx6dt) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L9)):*
> ```
>     [Write, Description("A combination of one or more of the ActiveDirectoryRights enumeration values that specifies the rights of the access rule. Default is 'GenericAll'."), ValueMap{"AccessSystemSecurity", "CreateChild", "Delete", "DeleteChild", "DeleteTree", "ExtendedRight", "GenericAll", "GenericExecute", "GenericRead", "GenericWrite", "ListChildren", "ListObject", "ReadControl", "ReadProperty", "Self", "Synchronize", "WriteDacl", "WriteOwner", "WriteProperty"}, Values{"AccessSystemSecurity", "CreateChild", "Delete", "DeleteChild", "DeleteTree", "ExtendedRight", "GenericAll", "GenericExecute", "GenericRead", "GenericWrite", "ListChildren", "ListObject", "ReadControl", "ReadProperty", "Self", "Synchronize", "WriteDacl", "WriteOwner", "WriteProperty"}] String ActiveDirectoryRights[];
>     [Key, Description("Indicates whether to Allow or Deny access to the target object."), ValueMap{"Allow", "Deny"}, Values{"Allow", "Deny"}] String AccessControlType;
>     [Key, Description("The schema GUID of the object to which the access rule applies.")] String ObjectType;
> ```

Wonder if we instead should name it 'ObjectSchemaGuid' so it is easier what is expected? 
Another option that would be even better is if we could assign it to the SchemaObjectName and the code looks up the Guid.  Would be much easier to read a configuration if the name was in the configuration instead of the Guid.

Maybe we could get the guid using something like this? 
```
$schemaNamingContext = (Get-ADRootDSE).SchemaNamingContext
[System.GUID] (Get-ADObject -SearchBase $schemaNamingContext -LDAPFilter "(LdapDisplayName=Computer)" -Properties SchemaIdGuid).SchemaIdGuid
```

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 10 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfZkky8n_CN-mGyN2E:-LBfZkky8n_CN-mGyN2F:beqh4kp) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L10)):*
> ```
> ActiveDirectorySecurityInheritance
> ```

maybe we just can call this Inheritance?

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 11 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfTsHm8IhzaCCURMMT:-LBfTsHm8IhzaCCURMMU:b-6bmq8g) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L11)):*
> ```
>     [Key, Description("The schema GUID of the object to which the access rule applies.")] String ObjectType;
>     [Key, Description("One of the 'ActiveDirectorySecurityInheritance' enumeration values that specifies the inheritance type of the access rule."), ValueMap{"All", "Children", "Descendents", "None", "SelfAndChildren"}, Values{"All", "Children", "Descendents", "None", "SelfAndChildren"}] String ActiveDirectorySecurityInheritance;
>     [Key, Description("The schema GUID of the child object type that can inherit this access rule.")] String InheritedObjectType;
> ```

Same as previous comment about the name. 

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 11 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfW_V867qvOsfxHR40:-LBfW_V867qvOsfxHR41:b-vnl30z) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L11)):*
> ```
>     [Key, Description("The schema GUID of the object to which the access rule applies.")] String ObjectType;
>     [Key, Description("One of the 'ActiveDirectorySecurityInheritance' enumeration values that specifies the inheritance type of the access rule."), ValueMap{"All", "Children", "Descendents", "None", "SelfAndChildren"}, Values{"All", "Children", "Descendents", "None", "SelfAndChildren"}] String ActiveDirectorySecurityInheritance;
>     [Key, Description("The schema GUID of the child object type that can inherit this access rule.")] String InheritedObjectType;
> ```

Must we have this as a key. What if the permission should not be able to be inherited? 

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/186)*
<!-- Sent from Reviewable.io -->

claudiospizzi commented:
Review status: 0 of 4 files reviewed at latest revision, 10 unresolved discussions.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 1 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfKBLM1iHXlhvJXCNi:-LCYYTwL3YTT3DjyPPNQ:bz91689) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L1)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> xADObjectPermissionEntry
> ```

Is it a special reason we should end the name with 'Entry'? Would it be misleading to call it just 'xADObjectPermission'?
</blockquote></details>

I've used the suffix "Entry" to specify, that this is just one entry in the list of premissions (ACL) for one AD object. It will only interact with the one permission and leave all others as they were.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 5 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfUqiLC-d21git-10n:-LCYZ58H3D25xQlu_4nb:bb6un1j) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L5)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Should we make the description more descriptive and say something about that this is the object (for example and OU?) that will get the permission?
</blockquote></details>

Done

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 6 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfKald4_65wNZ9w0SN:-LCYZo2183v4l-A96H5A:b-den2ae) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L6)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> IdentityReference
> ```

Is there a special reason to add 'Reference' here? Can it only be assigned either SamAccountName, CommonName, UserPrincipalName, DisplayName or is it possible to provide any them? Thinking if we should call it just 'Identity', or if we should use any existing name that for example xUser uses?  
</blockquote></details>

All resource properties are named like the System.DirectoryServices.ActiveDirectoryAccessRule object you will get with the Get-Acl command. The idea is, that someone can create the delegation by manually, then use Get-Acl to read them and apply them one-to-one to the DSC resource. 

`Get-Acl 'AD:\DC=contoso,DC=com' | % Access | select -First 1`

This will output somethink like this:
```
ActiveDirectoryRights : DeleteChild
InheritanceType       : None
ObjectType            : 00000000-0000-0000-0000-000000000000
InheritedObjectType   : 00000000-0000-0000-0000-000000000000
ObjectFlags           : None
AccessControlType     : Deny
IdentityReference     : Everyone
IsInherited           : False
InheritanceFlags      : None
PropagationFlags      : None
```

The IdentityReference is used to create a System.Security.Principal.NTAccount, but the constructor of this object does not specify what is valid. So I would say, just SamAccountName because that's what I've tested.
https://msdn.microsoft.com/en-us/library/31k4bah1

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 6 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfW4HU5tkO3HUvcEYm:-LCYaxCY0M3aSrpi6QVb:b-4sjcao) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L6)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We should say in the description that this can be only be set to an user or group account.
We should replace the abbreviation 'ace' with something that it is this that gets 'permission'.
</blockquote></details>

Replaced with 'permission entry'.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 7 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfLizg350lEka_feEi:-LCYb4J4EsOW2Gy7joOI:b2nreh6) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L7)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> ActiveDirectoryRights
> ```

Maybe call this property `Permission`?
</blockquote></details>

Same as above.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 9 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfRuZI7fGr7oUYo8fp:-LCYb5Yt7gYy_Y7v_hfV:b2nreh6) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L9)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Wonder if we instead should name it 'ObjectSchemaGuid' so it is easier what is expected? 
Another option that would be even better is if we could assign it to the SchemaObjectName and the code looks up the Guid.  Would be much easier to read a configuration if the name was in the configuration instead of the Guid.

Maybe we could get the guid using something like this? 
```
$schemaNamingContext = (Get-ADRootDSE).SchemaNamingContext
[System.GUID] (Get-ADObject -SearchBase $schemaNamingContext -LDAPFilter "(LdapDisplayName=Computer)" -Properties SchemaIdGuid).SchemaIdGuid
```
</blockquote></details>

Same as above.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 10 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfZkky8n_CN-mGyN2E:-LCYbFzp5kBw_oOXQjdx:b2nreh6) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L10)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> ActiveDirectorySecurityInheritance
> ```

maybe we just can call this Inheritance?
</blockquote></details>

Same as above.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 11 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfTsHm8IhzaCCURMMT:-LCYbGl-7Jcv1hpfp7KR:b2nreh6) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L11)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Same as previous comment about the name. 
</blockquote></details>

Same as above.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 11 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfW_V867qvOsfxHR40:-LCYbTdCBToezxn5lgvp:bvxy4cf) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L11)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Must we have this as a key. What if the permission should not be able to be inherited? 
</blockquote></details>

I think then we use a null guid '00000000-0000-0000-0000-000000000000'.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/186)*
<!-- Sent from Reviewable.io -->

claudiospizzi commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfIRi83t2cC3fETmem:-LCYenG90j3ll9ODNK0K:b5078jp) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

I started reviewing this but soon realized that the Key's in the schema seems that they could become a problem in certain scenarios (I can be wrong). For me to get a bigger understanding for this resource, could you provide a few examples?
Currently there are no Examples folder, so maybe we can do something similar to https://github.com/PowerShell/SqlServerDsc/tree/dev/Examples? 

Can we start with three examples.

1. Give a prestaged VCO full control permissions (aka GenericAll) to the CNO.
2. Allow a group permission to create and delete  ('CreateChild,DeleteChild') computer objects  in an OU and all sub-OUs that may get created.
3. Allow a group permission to read and write ('ReadProperty,WriteProperty') all properties of computer objects in an OU and all sub-OUs that may get created


</blockquote></details>

I've added the examples:
https://github.com/claudiospizzi/xActiveDirectory/tree/xADObjectPermissionEntry/Examples

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/186)*
<!-- Sent from Reviewable.io -->

claudiospizzi commented:
Review status: 0 of 8 files reviewed at latest revision, 10 unresolved discussions.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 6 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfKald4_65wNZ9w0SN:-LCYfhqh7XxaqRJlL_uj:b255qkl) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L6)):*
<details><summary><i>Previously, claudiospizzi (Claudio Spizzi) wrote…</i></summary><blockquote>

All resource properties are named like the System.DirectoryServices.ActiveDirectoryAccessRule object you will get with the Get-Acl command. The idea is, that someone can create the delegation by manually, then use Get-Acl to read them and apply them one-to-one to the DSC resource. 

`Get-Acl 'AD:\DC=contoso,DC=com' | % Access | select -First 1`

This will output somethink like this:
```
ActiveDirectoryRights : DeleteChild
InheritanceType       : None
ObjectType            : 00000000-0000-0000-0000-000000000000
InheritedObjectType   : 00000000-0000-0000-0000-000000000000
ObjectFlags           : None
AccessControlType     : Deny
IdentityReference     : Everyone
IsInherited           : False
InheritanceFlags      : None
PropagationFlags      : None
```

The IdentityReference is used to create a System.Security.Principal.NTAccount, but the constructor of this object does not specify what is valid. So I would say, just SamAccountName because that's what I've tested.
https://msdn.microsoft.com/en-us/library/31k4bah1
</blockquote></details>

The only property which does not match is `ActiveDirectorySecurityInheritance` to `InheritanceType` in PowerShell. This matches with the Enumeration type:
https://msdn.microsoft.com/en-us/library/w72e8e69

We can change this, so that it's matching one-to-one with the PowerShell Get-Acl cmdlet.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/186)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 1 of 7 files at r4.
Review status: 1 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfIRi83t2cC3fETmem:-LCnegaT7rky5N4WXx6Q:buzf0bu) (no related file):*
<details><summary><i>Previously, claudiospizzi (Claudio Spizzi) wrote…</i></summary><blockquote>

I've added the examples:
https://github.com/claudiospizzi/xActiveDirectory/tree/xADObjectPermissionEntry/Examples
</blockquote></details>

Thanks for adding the examples! I didn't have time to review this fully yet. I will get back to it this weekend or on Monday! Sorry!

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 1 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfKBLM1iHXlhvJXCNi:-LCnaCnT3f7Ci2_fwFvm:b7lujm8) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L1)):*
<details><summary><i>Previously, claudiospizzi (Claudio Spizzi) wrote…</i></summary><blockquote>

I've used the suffix "Entry" to specify, that this is just one entry in the list of premissions (ACL) for one AD object. It will only interact with the one permission and leave all others as they were.
</blockquote></details>

Okay! Maybe we could add the to the resource description?

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 6 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfKald4_65wNZ9w0SN:-LCnaw5XAPRr8RHzpgCJ:b-6ogyu) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L6)):*
<details><summary><i>Previously, claudiospizzi (Claudio Spizzi) wrote…</i></summary><blockquote>

The only property which does not match is `ActiveDirectorySecurityInheritance` to `InheritanceType` in PowerShell. This matches with the Enumeration type:
https://msdn.microsoft.com/en-us/library/w72e8e69

We can change this, so that it's matching one-to-one with the PowerShell Get-Acl cmdlet.
</blockquote></details>

I rather see we are consistent with the naming withing the resource module than what an individual cmdlet returns, and that the user understand what value goes into the parameter. But with that said I will leave it to you to decide the best names here. 

If you only tested this with SamAccountName maybe we should add that to the parameter description so the user doesn't add a value that does not work.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 9 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfRuZI7fGr7oUYo8fp:-LCndcJe9pVyELkAM4-E:b-hn81xv) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L9)):*
<details><summary><i>Previously, claudiospizzi (Claudio Spizzi) wrote…</i></summary><blockquote>

Same as above.
</blockquote></details>

The problem I see here is that users probably won't add the permission manually first and then build a configuration from that. They most likely want to be able to create the configuration from scratch and the apply the configuration to a target node. I think we should make the process easier than asking the user to look up guids? But that might not be possible?

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 11 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfW_V867qvOsfxHR40:-LCneVVw6VJtHAPQRUus:b-7ksq2n) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L11)):*
<details><summary><i>Previously, claudiospizzi (Claudio Spizzi) wrote…</i></summary><blockquote>

I think then we use a null guid '00000000-0000-0000-0000-000000000000'.
</blockquote></details>

We should document that in the parameter description(s). 

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/186)*
<!-- Sent from Reviewable.io -->

johlju commented:
@claudiospizzi I have not forgotten this! It will take half a day to review this one, so I just haven't had the time yet. But will get to this!

johlju commented:
@claudiospizzi sorry it took so long for me to get to this one! I ran you examples through my lab to verify. Looks promising! Awesome work on this one, clean code! :smile: 

Just need the review comments resolved/commented on and then we can merge this one. Maybe we can get it merged before next week, then we will be able to get this one in the next release.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfIRi83t2cC3fETmem:-LEANYMw5bZLjaU8uMgv:b-ic7iel) (no related file):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Thanks for adding the examples! I didn't have time to review this fully yet. I will get back to it this weekend or on Monday! Sorry!
</blockquote></details>

Examples LGTM (just minor comment).

---

*[README.md, line 213 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAO6TsCfgaQHPmAdjZ:-LEAO6TsCfgaQHPmAdj_:b-20lv4a) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/README.md#L213)):*
> ```
> objects-
> ```

Think it should be a `.` at the end there instead?

---

*[README.md, line 220 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEANjrP574jQhLMW2a0:-LEANjrP574jQhLMW2a1:br6n5vh) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/README.md#L220)):*
> ```Markdown
> * **ActiveDirectoryRights**: A combination of one or more of the ActiveDirectoryRights enumeration values that specifies the rights of the access rule. Default is 'GenericAll'. Valid values:
>   * AccessSystemSecurity
>   * CreateChild
> ```

We should add these as possible values as `{ AccessSystemSecurity | CreateChild | ..etc }` at the end of the description.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1, line 92 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAObvRDCUzBMXF7nuV:-LEAObvRDCUzBMXF7nuW:b-mgtwv4) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1#L92)):*
> ```PowerShell
>                 $access.InheritedObjectType.Guid -eq $InheritedObjectType)
>             {
>                 $returnValue['Ensure'] = 'Present'
> ```

Maybe we can return a verbose message saying that it was found or not found here?

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1, line 202 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAP8YY5nNTOFgLsoQw:-LEAP8YY5nNTOFgLsoQx:b-pp44vp) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1#L202)):*
> ```PowerShell
>     }
> 
>     if ($Ensure -eq 'Absent')
> ```

Maybe we should use `else` here instead so that it's clear it can never be both 'Present' and 'Absent' at the same time?

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1, line 308 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAPPwr5L8FupnlnMjw:-LEAPPwr5L8FupnlnMjx:b-jlj7uk) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1#L308)):*
> ```PowerShell
> 
>     # Get the current state
>     $currentState = Get-TargetResource -Path $Path -IdentityReference $IdentityReference -AccessControlType $AccessControlType -ObjectType $ObjectType -ActiveDirectorySecurityInheritance $ActiveDirectorySecurityInheritance -InheritedObjectType $InheritedObjectType
> ```

Can we use splatting for these parameters instead? Make the row less long. :)

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1, line 313 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAPa6wDUvqaGh41bqW:-LEAPa6wDUvqaGh41bqX:b-zcbm90) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1#L313)):*
> ```
> checn
> ```

check

---

*[Examples/Resources/xADObjectPermissionEntry/1-DelegateFullControl.ps1, line 3 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAD-R55-S8ExJbItvm:-LEAD-R55-S8ExJbItvn:b-yxit2u) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/Examples/Resources/xADObjectPermissionEntry/1-DelegateFullControl.ps1#L3)):*
> ```
> GeneralAll
> ```

GenericAll

---

*[Examples/Resources/xADObjectPermissionEntry/1-DelegateFullControl.ps1, line 3 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAD6Dp-gRmANchEVSS:-LEA6t_FBbOycY9tMNZ6:bxt6apj) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/Examples/Resources/xADObjectPermissionEntry/1-DelegateFullControl.ps1#L3)):*
> ```
>         In this example, we will add full control (GeneralAll) permissions to
>         the virtual computer object (VCO) for a cluster name object (CNO).
> ```

Can we add which is VCO and which is CNO to easier read the example, or rename ROLE01 - it can me misread as a group?

---

*[Tests/Unit/MSFT_xADObjectPermissionEntry.Tests.ps1, line 90 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAPm0r6Fg1EjhJkDUf:-LEAPm0r6Fg1EjhJkDUg:b-ecy1av) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/Tests/Unit/MSFT_xADObjectPermissionEntry.Tests.ps1#L90)):*
> ```
> Context 'The desired ace is present'
> ```

We should start the descriptions with 'When...' on context-blocks.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/186#-:-LEAQCQr2nN9KbWXI_uR:b177p14)*
<!-- Sent from Reviewable.io -->

claudiospizzi commented:
Review status: 7 of 8 files reviewed, 13 unresolved discussions (waiting on @johlju and @claudiospizzi)

---

*[README.md, line 213 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAO6TsCfgaQHPmAdjZ:-LFURlXcCXr2GpJrBzQ2:bb6un1j) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/README.md#L213)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> objects-
> ```

Think it should be a `.` at the end there instead?
</blockquote></details>

Done

---

*[README.md, line 220 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEANjrP574jQhLMW2a0:-LFURkl-DNa0CQ2vdNz_:bb6un1j) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/README.md#L220)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We should add these as possible values as `{ AccessSystemSecurity | CreateChild | ..etc }` at the end of the description.
</blockquote></details>

Done

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1, line 92 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAObvRDCUzBMXF7nuV:-LFUS42JEJrUugrwDmVH:bb6un1j) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1#L92)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Maybe we can return a verbose message saying that it was found or not found here?
</blockquote></details>

Done

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1, line 202 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAP8YY5nNTOFgLsoQw:-LFUSA202H7-hAs9k0rI:b34o1dx) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1#L202)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Maybe we should use `else` here instead so that it's clear it can never be both 'Present' and 'Absent' at the same time?
</blockquote></details>

Sure, makes sence. Done.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1, line 308 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAPPwr5L8FupnlnMjw:-LFUSR_b-6WYh_CsGIeF:b4b9yoy) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1#L308)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Can we use splatting for these parameters instead? Make the row less long. :)
</blockquote></details>

Checked

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1, line 313 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAPa6wDUvqaGh41bqW:-LFUSU0r8UT3mfX8aTDE:bb6un1j) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.psm1#L313)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> checn
> ```

check
</blockquote></details>

Done

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 1 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfKBLM1iHXlhvJXCNi:-LFUStvlBVa_jDZS2oQ8:b-mywj9f) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L1)):*
> I've used the suffix "Entry" to specify, that this is just one entry in the list of premissions (ACL) for one AD object. It will only interact with the one permission and leave all others as they were.
Added

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 6 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfKald4_65wNZ9w0SN:-LFUTS8x7YtmR3D6J2TJ:b-4ahyp6) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L6)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

I rather see we are consistent with the naming withing the resource module than what an individual cmdlet returns, and that the user understand what value goes into the parameter. But with that said I will leave it to you to decide the best names here. 

If you only tested this with SamAccountName maybe we should add that to the parameter description so the user doesn't add a value that does not work.
</blockquote></details>

I will leave the property as it is. Comment added that the SamAccountName should be used for the IdentityReference.

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 9 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfRuZI7fGr7oUYo8fp:-LFUT_lv-lOmXV2ILspY:bkrykuw) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L9)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

The problem I see here is that users probably won't add the permission manually first and then build a configuration from that. They most likely want to be able to create the configuration from scratch and the apply the configuration to a target node. I think we should make the process easier than asking the user to look up guids? But that might not be possible?
</blockquote></details>

Yeah, I understand that. Maybe we can list the most common GUIDs in the readme or somethink like that. But I think, thats hard to decide. Maybe a command to get all possible GUIDs?

---

*[Examples/Resources/xADObjectPermissionEntry/1-DelegateFullControl.ps1, line 3 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAD6Dp-gRmANchEVSS:-LFUUGvYDKfi1dgwpsFs:b-5ghr5g) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/Examples/Resources/xADObjectPermissionEntry/1-DelegateFullControl.ps1#L3)):*
> GeneralAll
Done

---

*[Examples/Resources/xADObjectPermissionEntry/1-DelegateFullControl.ps1, line 3 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAD-R55-S8ExJbItvm:-LFUUIkG32xciD-PALSh:bb6un1j) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/Examples/Resources/xADObjectPermissionEntry/1-DelegateFullControl.ps1#L3)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

> ```
> GeneralAll
> ```

GenericAll
</blockquote></details>

Done

---

*[Tests/Unit/MSFT_xADObjectPermissionEntry.Tests.ps1, line 90 at r4](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LEAPm0r6Fg1EjhJkDUf:-LFUUl6I8lzXNSdTNs7b:b-x4kgl) ([raw file](https://github.com/powershell/xactivedirectory/blob/af5bc4b0cbc15e53e828c0cbce98d701355abcaf/Tests/Unit/MSFT_xADObjectPermissionEntry.Tests.ps1#L90)):*
> Context 'The desired ace is present'
Done

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/186)*
<!-- Sent from Reviewable.io -->

johlju commented:
Reviewed 5 of 5 files at r5.
Review status: all files reviewed, 4 unresolved discussions (waiting on @claudiospizzi and @johlju)

---

*[DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof, line 9 at r3](https://reviewable.io/reviews/powershell/xactivedirectory/186#-LBfRuZI7fGr7oUYo8fp:-LFWM6FV361qOzTcAykl:bwpqxie) ([raw file](https://github.com/powershell/xactivedirectory/blob/715aa668b7a9e2c0970d130ee5b1969f633391bf/DSCResources/MSFT_xADObjectPermissionEntry/MSFT_xADObjectPermissionEntry.schema.mof#L9)):*
<details><summary><i>Previously, claudiospizzi (Claudio Spizzi) wrote…</i></summary><blockquote>

Yeah, I understand that. Maybe we can list the most common GUIDs in the readme or somethink like that. But I think, thats hard to decide. Maybe a command to get all possible GUIDs?
</blockquote></details>

A command to get all possible guids would help, good idea. Should we add that to the resource description? Or link to a page explaining the guid's?  The  '00000000-0000-0000-0000-000000000000' should be explained though, see next ciomment.

---


*Comments from [Reviewable](https://reviewable.io/reviews/powershell/xactivedirectory/186)*
<!-- Sent from Reviewable.io -->

claudiospizzi commented:
Hi @johlju 
I've updated the module to fix the mof issue and added additional description. You can review the latest changes again? Would be great if it make into the August DSC Resource Kit Release July.

johlju commented:
The command to get all possible GUIDs can be an enhancement for a later PR.
